File: JavaScript/firstUnique.js
// Given a string, find the first non-repeating character in it and return it's index. If it doesn't exist, return -1.
// Hash table, Works
function firstUniqChar(s) {
    
    let ht = {};
    
    for (let i = 0; i < s.length; i++) {
        ht[s[i]] ? ht[s[i]] += 1 : ht[s[i]] = 1;
    }
    
    
    for (let j = 0; j < s.length; j++) {
        if (ht[s.charAt(j)] === 1) {
            return j;
        }
    }
    return -1;
}
// Hash table, Works
function firstUniqChar(s) {
    
    let ht = {};
    
    for (let i = 0; i < s.length; i++) {
        ht[s[i]] = (ht[s[i]] || 0) + 1;
    }
    
    for (let j = 0; j < s.length; j++) {
        if (ht[s.charAt(j)] === 1) {
            return j;
        }
    }
    return -1;
}
// Array with ascii codes
function firstUniqChar(s) {
    
    let arr = [];
    
    for (let i = 0; i < s.length; i++) {
        if (arr[s.charCodeAt(i)]) {
            arr[s.charCodeAt(i)] += 1;
        }
        else {
        arr[s.charCodeAt(i)] = 1;
        }
    }
    
    for (let i = 0; i < s.length; i++) {
        if (arr[s.charCodeAt(i)] === 1) {
            return i;
        }
    }
    return -1;
}

File: C-Sharp/FirstUnique.cs
// Given a string, find the first non-repeating character in it and return it's index. If it doesn't exist, return -1.
// Works
public class Solution {
    public int FirstUniqChar(string s) {
        Dictionary<char, int> ht = new Dictionary<char, int>();
        
        for (int i = 0; i < s.Length; i++)
        {
            if (!ht.ContainsKey(s[i]))
            {
                ht[s[i]] = 1;
            }
            else
            {
                ht[s[i]] += 1;
            }
        }
        for (int i = 0; i < s.Length; i++)
        {
            if (ht[s[i]] == 1)
            {
                return i;
            }
        }
        return -1;
    }
}
// Using array, faster!
public class Solution {
    public int FirstUniqChar(string s) {
        int[] arr = new int[128];
        for (int i = 0; i < s.Length; i++)
        {
            arr[s[i]]++;
        }
        for (int i = 0; i < s.Length; i++)
        {
            if (arr[s[i]] == 1)
            {
                return i;
            }
        }
        return -1;
    }
}